bossplot:  add plot range option

dcunning11235 opened the pull request:
Added --plot-range MIN MAX option to bossplot; altered flux axis scaling code
    to handle properly scaling within a specified range.
Added a couple of sentences to bin/scripts.rst RE: the new option as well as
    several of the existing options that weren't yet mentioned.

dcunning11235 committed:
Added --plot-range MIN MAX option to bossplot; altered flux axis scaling code
	to handle properly scaling only within a specified range.
Added a couple of sentences to bin/scripts.rst RE: the new option as well as
	several of the existing options that weren't yet mentioned.

dkirkby commented:
Relates to #27

dkirkby commented on bin/scripts.rst:
```diff
@@ -112,6 +112,10 @@ This should open a new window containing the plot that you will need to close in
 
 You can also save the data shown in a plot using ``--save-data`` with an optional filename (the default is ``bossplot-{plate}-{mjd}-{fiber}.dat``).  Data is saved using the `ascii.basic <http://docs.astropy.org/en/latest/api/astropy.io.ascii.Basic.html#astropy.io.ascii.Basic>`_ format and only wavelengths with valid data are included in the output.
 
+Use ``--plot-range MIN MAX`` to specify a wavelenght range over which to plot, overriding the default, auto-detected range.  ``--scatter`` will give a scatter plot of the flux rather than a the flux +/- error band.
```
typo: wavelength

dcunning11235 committed:
Corrected typos in bin/scripts.rst

dkirkby commented:
Since you have opened this can of worms...

There are 3 axes that we potentially want to be able to customize from the command line:
- x-axis, which is always wavelength. The units are currently always Angstroms but we might also want to support co-add pixel units = 1e-4*log10(lambda).
- left y-axis, which is usually flux, but could be in electrons.
- optional right y-axis, which is currently always dispersion in co-add pixel units.

Currently, the x-axis is set to show the full range of bins with valid data (according to the numpy mask), the left y-axis shows the 0.5% - 99.5% quantile, and the optional right y-axis uses the matplot lib default scaling.  There are no cmd-line args to customize any of this yet, since this was forseen for issue #27.

We want a consistent way to handle all these via cmd-line args, so instead of `--plot-range` for the x-axis, we probably want arg names like  `--wlen-*`, `--flux-*`, `--wdisp-*`. For each axis, there is a min and a max value.  Do we use separate cmd-line args for these:

```
--wlen-min ? --wlen-max ?
```

or combine them:

```
--wlen-range ?:?
```

In either case, it should be possible to only set one limit and use the default for the other.  Finally, the vertical ranges should support either absolute or percentile limits. (Percentile limits don't make much sense for the x-axis, but its a good idea to keep the API consistent anyway).

dcunning11235 commented:
That is a can of worms.

Okay, I can do that.  I'm closing the pull request until all changes pushed.

dkirkby commented:
@dcunning11235 Its not a bad idea to open a pull request early on while implementing a new feature: that allows others to give feedback earlier in the development process and ensures that regression tests are run as you add commits.  When working this way, you keep the PR assigned to yourself until you are ready for a final review, when you assign it to someone else.

dcunning11235 committed:
Added in --wlen-range (was --plot-range) and updated it to take
python-eqsue ranges.  Ranges can be either absolute values,
percentages, or blank (for default, auto value); these may be mixed.
E.g.: --wlen-range 4000:8000
	--wlen-range 4000:
	--wlen-range :90%
	--wlen-range 20%:80%
	--wlen-range 10%:5000
are all valid entries.

Added in new function "abs_or_pct_range" used as type in
parser.add_argument add call to handle the above; this results in
namedtuple Range value for this option.

dcunning11235 commented:
@dkirkby Okay, makes sense.

First checkin done, just with support for --wlen-range (replacing --plot-range)

dkirkby commented:
Do you prefer `--wlen-range` to `--wlen-min` and `--wlen-max` ?  I think it is probably more user friendly, but also requires more code to handle cases where only one limit is specified.  Of course, the code for turning cmd-line args into numerical ranges can (and should) be re-used for all 3 axes.

dcunning11235 commented:
I think having the range is better; the less typing the better, nearly always.  (And it's already implemented :>)

Yup, the method (and tuples) for handling the range from the command(s) are already generic.  The code for actually applying it to wavelength can be cleaned up a bit more, I think; when I go back and look at that I will see what can get pulled out and reused.

dcunning11235 committed:
Added most of support for --flux-range, but not fully tested; has
issue when combined with --wlen-range.

Corrected an issue in argparse messaging string (% are escape chars,
turns out.)

Updated --flux-range messaging re: passing negative values through
argparse.

dcunning11235 committed:
Now fully supporting --wlen-range and --flux-range.

Cleaned up implementation so that code is now being reused; made a pass
at making generic.

dcunning11235 committed:
Fininshed, with --wdisp-range now added in

dcunning11235 committed:
Updated scripts.rst to document new bossplot options

dcunning11235 committed:
Multiple small tweaks, bug fixes, etc.  Clean up some code (without changing functionality) and cleanup of comments.

Ready for testing.

dcunning11235 commented:
I think this should be ready; throw the sink at it.

dcunning11235 commented:
Assigning back to self, need to look at changes made by @dmargala

dcunning11235 commented:
Superseded by https://github.com/dkirkby/bossdata/pull/59; this should not be merged.
